### DISCUSSION TITLE
How can I get type definitions in VS Code intellisense?

### DISCUSSION THREAD

**Discussion by nthouliss** [supabase/supabase#12426] [Questions] [answered]
We're migrating to Supabase V2 and have injected the definitions as described in the [documentation](https://supabase.com/docs/reference/javascript/select#injecting-type-definitions) and we're not getting the type definitions at all, for example all the return types are `any`.

If we inject into the from statement (like what was done in V1) we manage to get the return type expected but none of the other auto-completion features (like what columns we can select from).

How exactly is the Typescript support expected to work in V2? Should it magically work once the definitions have been injected with the `createClient` function or are additional configurations expected?

Here's a minimal example that doesn't work:

```
export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json }
  | Json[];

export interface Database {
  public: {
    Tables: {
      user_assets: {
        Row: {
          id: string;
          name: string;
          owner: string;
        };
        Insert: {
          id: string;
          name: string;
          owner: string;
        };
        Update: {
          id?: string;
          name?: string;
          owner?: string;
        };
      };
    };
    Functions: {};
  };
}
```
<img width="1135" alt="Screenshot_2023-02-15_at_10 42 50_am" src="https://user-images.githubusercontent.com/53867736/218888194-eb7c58e0-2aa9-420b-9b32-cc203be593c5.png">

<img width="923" alt="Screenshot_2023-02-15_at_10 45 56_am" src="https://user-images.githubusercontent.com/53867736/218888438-e32dd5b6-06e4-4524-a997-ecbd0f283560.png">

Manually injecting into `from`:
![Screenshot_2023-02-15_at_10 55 23_am](https://user-images.githubusercontent.com/53867736/218890216-ba125427-f6e7-4570-a4c2-1813b9d89029.png)

**Comment by silentworks** [comment]
You are doing it the correct way by using the type at the point of creating the client. How are you creating the types that you pass to `createClient`?

**Reply by nthouliss** [reply]
The types were generated as per the [docs on generating types](https://supabase.com/docs/reference/javascript/typescript-support#generating-types) but it looks like that isn't generating a Views object in the interface and the supabase client errors because it expects it, so I guess there's a bug there but not sure where to report. 

For clarity, I changed the Database interface to this to get it working:
```
export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json }
  | Json[];

export interface Database {
  public: {
    Tables: {
      user_assets: {
        Row: {
          id: string;
          name: string;
          owner: string;
        };
        Insert: {
          id: string;
          name: string;
          owner: string;
        };
        Update: {
          id?: string;
          name?: string;
          owner?: string;
        };
      };
    };
    Functions: {};
    Views: {};
  };
}

**Comment by nthouliss** [comment, marked as answer]
Here's the answer in case anyone else needs it. You need to make sure your interface has Tables, Functions, and Views defined.

```
export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json }
  | Json[];

export interface Database {
  public: {
    Tables: {};
    Functions: {};
    Views: {};
  };
}
```
